shim/shim.c

```diff
@@ -0,0 +1,292 @@
+
+/* globals */
+/* Pipe used for capturing signal occurence */
+int pipe_fd[2];
+
+void add_pollfd(struct pollfd *poll_fds, nfds_t *nfds, int fd,  short events) {
+	struct pollfd pfd = { 0 };
+
+	assert(*nfds < MAX_POLL_FDS);
+	pfd.fd = fd;
+	pfd.events = events;
+	poll_fds[*nfds] = pfd;
+	(*nfds)++;
+}
+
+static void
+signal_handler(int signal_no)
+{
+	int savedErrno;                     /* In case we change 'errno' */
+ 
+	savedErrno = errno;
+	/* Write signal number to pipe, so that the signal can be identfied later */
+	if (write(pipe_fd[1], &signal_no, sizeof(signal_no)) == -1 && errno != EAGAIN) {
+		return;
+	}
+	errno = savedErrno;
+}
+
+void
+err_exit(const char *format, ...)
+{
+	va_list	args;
+
+	va_start(args, format);
+	vfprintf(stderr, format, args);
+	va_end(args);
+	exit(EXIT_FAILURE);
+}
+
+void
+set_fd_nonblocking(int fd)
+{
+	int flags;
+
+	flags = fcntl(fd, F_GETFL);
+	if (flags == -1) {
+		err_exit("fcntl-F_GETFL");
+	}
+	flags |= O_NONBLOCK;
+
+	if (fcntl(fd, F_SETFL, flags) == -1) {
+		err_exit("Error setting fd as nonblocking: %s\n", strerror(errno));
+	}
+}
+
+void
+set_big_endian_32(uint8_t *buf, uint32_t val)
+{
+        buf[0] = (uint8_t)(val >> 24);
+        buf[1] = (uint8_t)(val >> 16);
+        buf[2] = (uint8_t)(val >> 8);
+        buf[3] = (uint8_t)val;
+}
+
+uint32_t get_big_endian_32(char *buf) {
+	return (uint32_t)(buf[0] >> 24 | buf[1] >> 16 | buf[2] >> 8 | buf[3]);
+}
+
+char*
+get_hypertart_msg(char *json, int hyper_cmd_type, size_t *len) {
+	char *hyperstart_msg = NULL;
+
+	*len = strlen(json) + 8 + 1;
+	hyperstart_msg = malloc(sizeof(char) * *len);
+	if (! hyperstart_msg) {
+		abort();
+	}
+	set_big_endian_32((uint8_t*)hyperstart_msg, (uint32_t)hyper_cmd_type);
+	set_big_endian_32((uint8_t*)hyperstart_msg+4, (uint32_t)(strlen(json) + 8));
+	strcpy(hyperstart_msg+8, json);
+
+	return hyperstart_msg;
+ }
+
+void
+send_proxy_hyper_message(int fd, int hyper_cmd_type, char *json) {
+	char *proxy_hyper_msg = NULL;
+	size_t len, offset = 0;
+	int ret;
+	char *id = 1;
+	uuid_t uuid;
+	char *proxy_command = "hyper";
+
+	ret = asprintf(&proxy_hyper_msg,
+			"{\"command\":\"%s\",\"data\":\"%s\"}",
+			proxy_command, json);
+
+	if (ret == -1) {
+		abort();
+	}
+
+	len = strlen(proxy_hyper_msg + 1);
+
+	while (offset < len) {
+		ret = write(fd, proxy_hyper_msg + offset, len-offset);
+		if (ret == EINTR) {
+			continue;
+		}
+		if (ret <= 0 ) {
+			free(proxy_hyper_msg);
+			return;
+		}
+		offset += ret;
+	}
+	free(proxy_hyper_msg);
+}
+
+void
+handle_signals(char *container_id, int outfd) {
+	int sig;
+	char *buf;
+	int ret;
+	int cmd_type;
+	struct winsize ws;
+ 
+	while (read(pipe_fd[0], &sig, sizeof(sig)) != -1) {
+		if (sig == SIGWINCH ) {
+			cmd_type = WINSIZE;
+			if (ioctl(STDIN_FILENO, TIOCGWINSZ, &ws) == -1) {
+				fprintf(stderr, "Error getting the current window size: %s\n",
+					strerror(errno));
+				continue;
+			}
+			ret = asprintf(&buf, "{\"container_id\":\"%s\", \"row\":\"%d\", \"col\":\"%d\"}",
```

Comment:
Might be useful to have a debug option so that this function can log the action performed:

- `"killed container %s with signal %d"`
- `"handled SIGWINCH for container %s (row=%d, col=%d)"`